databricks_cli/pipelines/api.py

```diff
@@ -0,0 +1,182 @@
+        self.dbfs_client = DbfsApi(api_client)
+
+    def deploy(self, spec, headers=None):
+        lib_objects = LibraryObject.from_json(spec.get('libraries', []))
+        local_lib_objects, external_lib_objects = \
+            self._identify_local_libraries(lib_objects)
+
+        spec['libraries'] = LibraryObject.to_json(external_lib_objects +
+                                                  self._upload_local_libraries(local_lib_objects))
+        spec['credentials'] = self._get_credentials_for_request()
+        self.client.client.perform_query('PUT',
+                                         '/pipelines/{}'.format(spec['id']),
+                                         data=spec,
+                                         headers=headers)
+
+    def delete(self, pipeline_id, headers=None):
+        self.client.delete(pipeline_id, self._get_credentials_for_request(), headers)
+
+    @staticmethod
+    def _identify_local_libraries(lib_objects):
+        """
+        Partitions the given set of libraries into local and those already present in dbfs/s3 etc.
+        Local libraries are (currently) jar files with a file scheme or no scheme at all.
+        All other libraries should be present in a supported external source.
+        :param lib_objects: List[LibraryObject]
+        :return: List[List[LibraryObject], List[LibraryObject]] ([Local, External])
+        """
+        local_lib_objects, external_lib_objects = [], []
+        for lib_object in lib_objects:
+            parsed_uri = urllib.parse.urlparse(lib_object.path)
+            if lib_object.lib_type == 'jar' and parsed_uri.scheme == '':
+                local_lib_objects.append(lib_object)
+            elif lib_object.lib_type == 'jar' and parsed_uri.scheme.lower() == 'file':
+                # exactly 1 or 3
+                if parsed_uri.path.startswith('//') or parsed_uri.netloc != '':
+                    raise RuntimeError('invalid file uri scheme, '
+                                       'did you mean to use file:/ or file:///')
+                local_lib_objects.append(LibraryObject(lib_object.lib_type, parsed_uri.path))
+            else:
+                external_lib_objects.append(lib_object)
+        return local_lib_objects, external_lib_objects
+
+    def _upload_local_libraries(self, local_lib_objects):
+        remote_lib_objects = [LibraryObject(llo.lib_type, self._get_hashed_path(llo.path))
+                              for llo in local_lib_objects]
+
+        transformed_remote_lib_objects = [LibraryObject(rlo.lib_type, DbfsPath(rlo.path))
+                                          for rlo in remote_lib_objects]
+        upload_files = [llo_tuple for llo_tuple in
+                        zip(local_lib_objects, transformed_remote_lib_objects)
+                        if not self.dbfs_client.file_exists(llo_tuple[1].path)]
+
+        for llo, rlo in upload_files:
+            try:
+                self.dbfs_client.put_file(llo.path, rlo.path, False)
```

Comment:
Won't this fail your test case below with the same file contents?

Reply:
This function is mocked in the test